gpytorch/lazy/diag_lazy_tensor.py

```diff
@@ -30,41 +26,15 @@ def __add__(self, other):
 
         return AddedDiagLazyTensor(other, self)
 
-    def _get_indices(self, left_indices, right_indices, *batch_indices):
-        """Extract elements from the LazyTensor. Supports arbitrary batch sizes.
+    def _expand_batch(self, batch_shape):
+        return self.__class__(self._diag.expand(*batch_shape, self._diag.size(-1)))
 
-        Args:
-            :attr:`left_indices` (LongTensor): An `i`-dim tensor of row indices (these are the
-                same across all batches).
-            :attr:`left_indices` (LongTensor): An `i`-dim tensor of column indices (these are the
-                same across all batches).
-            :attr:`batch_indices` (LongTensor): A variable number of `i`-dim tensors of batch indices,
-                one for each batch dimension. If smaller than the number of batch dimensions of the
-                DiagLazyTensor, select all elements of the omitted batches.
-
-        Returns:
-            An `i`-dim tensor containing the requested elements if all batch index tensors were
-            passed in, otherwise a `bl x ... bk x i` batch of such tensors, where `l = len(batch_indices) + 1`.
-
-        """
-        batch_dims = len(batch_indices)
-        if batch_dims > self.batch_dim:
-            raise RuntimeError(
-                "Received {} batch index tensors, DiagLazyTensor has {} batches".format(batch_dims, self.batch_dim)
-            )
-        full_batch_shape = self._diag.shape[batch_dims:-1]
-        if len(full_batch_shape) == 0:
-            diag_elements = self._diag[batch_indices + (left_indices,)]
-        else:
-            d = torch.ones_like(left_indices)
-            diag_elements = torch.stack(
-                [
-                    self._diag[batch_indices + tuple(j * d for j in js) + (left_indices,)]
-                    for js in product(*[range(b) for b in full_batch_shape])
-                ]
-            )
-        equal_indices = (left_indices == right_indices).type_as(self._diag)
-        return diag_elements * equal_indices
+    def _get_indices(self, row_index, col_index, *batch_indices):
+        res = self._diag[(*batch_indices, row_index)]
+        # If row and col index don't agree, then we have off diagonal elements
+        # Those should be zero'd out
+        res = res * torch.eq(row_index, col_index).type_as(res)
```

Comment:
`type_as` will not honor the device id on cuda. Use `.to(device=res.device, dtype=res.dtype)` instead.